Bugfix / Fix issues with "scan all channels"

itayx opened the pull request:
As part of the new "Scan on all channels" the client can requests a scan
with a specific pool what will indicate they wish to run a scan on all
the radio's supported channels.
A channel pool with only a single channel with a value of '0' will
indicate that the client wants to use the "Scan on all channels".

kantera800 commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2298,6 +2298,11 @@ int db::get_channel_scan_dwell_time_msec(const sMacAddr &mac, bool single_scan)
 bool db::is_channel_scan_pool_supported(const sMacAddr &mac,
                                         const std::unordered_set<uint8_t> &channel_pool)
 {
+    if (channel_pool.size() == 1 && channel_pool.find(0) != channel_pool.end()) {
```
you should use a define 

```suggestion
    if (channel_pool.size() == 1 && *channel_pool.begin() == SCANALLCHAN) {
```

itayx replied:
Changed

kantera800 commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2359,6 +2364,12 @@ bool db::is_channel_in_pool(const sMacAddr &mac, uint8_t channel, bool single_sc
     auto &pool =
         (single_scan ? hostap->single_scan_config : hostap->continuous_scan_config).channel_pool;
+
+    if (pool.size() == 1 && pool.find(0) != pool.end()) {
```
same here

itayx replied:
Changed

kantera800 reviewed (approved):
Approved , 
just fix the condition

tomereli commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2298,6 +2298,11 @@ int db::get_channel_scan_dwell_time_msec(const sMacAddr &mac, bool single_scan)
 bool db::is_channel_scan_pool_supported(const sMacAddr &mac,
                                         const std::unordered_set<uint8_t> &channel_pool)
 {
+    if (channel_pool.size() == 1 && *channel_pool.begin() == SCAN_ALL_CHANNELS) {
```
I don't like this at all. If the user wants to scan all channels, then add an API to scan on all channels. APIs should be simple - this function based on its name should only check if the given pool is supported.

itayx replied:
See my comment below

tomereli commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2359,6 +2364,12 @@ bool db::is_channel_in_pool(const sMacAddr &mac, uint8_t channel, bool single_sc
     auto &pool =
         (single_scan ? hostap->single_scan_config : hostap->continuous_scan_config).channel_pool;
+
+    if (pool.size() == 1 && *pool.begin() == SCAN_ALL_CHANNELS) {
```
scan all channels should set the pool to empty and set a member bool scan_all_channels=true instead of this hack.

itayx replied:
See my comment below

itayx committed:
beerocks_defines: Add SCAN_ALL_CHANNELS define

Preemptive commit.
As part of the new "Scan on all channels" feature, a channel pool with
only a single channel with a value of '0' will indicate that the client
wants to use the "Scan on all channels"

Add SCAN_ALL_CHANNELS define

Signed-off-by: Itay Elenzweig <itayx.elenzweig@intel.com>

itayx committed:
controller: DCS task: revert changes

The old "Scan on all channels" worked from the DCS task.
This was not the best course of action as it failed to take into
consideration the "supported channel" validation when setting the pool
and the "result in pool" validation when the results are stored in the
DB.

Revert the changes made in 47d56372.

Signed-off-by: Itay Elenzweig <itayx.elenzweig@intel.com>

itayx commented:
@tomereli 
I don't think that creating a new API just for the "scan on all channels" is the correct course of action,
Instead I reworked the PR to do the following:
In the `set_channel_scan_pool` I expanded the signature to accept a "scan all channels" option (set by default to `false`) which, if set to `true` will fill the DCS channel pool with the radio's supported channels.
That way the DCS task will be unaware if the client uses the "scan all channels" feature, thus containing the changes only to the DB's `set_channel_scan_pool` function and its uses in the `son_management`.
I believe this is the better course of action since it allows the DB to the the sole manager of the DCS's channel pool, and we won't need to use "hacks" to fix the issues that rise from it

adam1985d reviewed (commented):
I think you meant to write `Preparative Commit` instead of `Preemptive commit`

adam1985d commented:
> @tomereli
> I don't think that creating a new API just for the "scan on all channels" is the correct course of action,
> Instead I reworked the PR to do the following:
> In the `set_channel_scan_pool` I expanded the signature to accept a "scan all channels" option (set by default to `false`) which, if set to `true` will fill the DCS channel pool with the radio's supported channels.
> That way the DCS task will be unaware if the client uses the "scan all channels" feature, thus containing the changes only to the DB's `set_channel_scan_pool` function and its uses in the `son_management`.
> I believe this is the better course of action since it allows the DB to the the sole manager of the DCS's channel pool, and we won't need to use "hacks" to fix the issues that rise from it

I have to say I disagree with the approach taken.
I think the son_management (not the task and not the DB) should handle the operation of `get supported channels` and then call the `set_channel_scan_pool` function with the channel_pool comprised of those channels.
This way you don't change the `set_channel_scan_pool` function at all and the `son_management` is handling the user request as it should.

itayx commented:
> > @tomereli
> > I don't think that creating a new API just for the "scan on all channels" is the correct course of action,
> > Instead I reworked the PR to do the following:
> > In the `set_channel_scan_pool` I expanded the signature to accept a "scan all channels" option (set by default to `false`) which, if set to `true` will fill the DCS channel pool with the radio's supported channels.
> > That way the DCS task will be unaware if the client uses the "scan all channels" feature, thus containing the changes only to the DB's `set_channel_scan_pool` function and its uses in the `son_management`.
> > I believe this is the better course of action since it allows the DB to the the sole manager of the DCS's channel pool, and we won't need to use "hacks" to fix the issues that rise from it
> 
> I have to say I disagree with the approach taken.
> I think the son_management (not the task and not the DB) should handle the operation of `get supported channels` and then call the `set_channel_scan_pool` function with the channel_pool comprised of those channels.
> This way you don't change the `set_channel_scan_pool` function at all and the `son_management` is handling the user request as it should.

I thought about it, but it will result in duplicated code for the single scan and the continuous scan.

tomereli commented on common/beerocks/bcl/include/bcl/beerocks_defines.h:
```diff
@@ -470,6 +470,7 @@ enum class eChannelScanOpErrCode : uint8_t {
 };
 
 #define CHANNEL_SCAN_INVALID_PARAM -1
+#define SCAN_ALL_CHANNELS 0
```
I really don't like this (did I say this already?). I still think that splitting APIs is better, but OK, we can do this in a separate PR I guess.

tomereli commented on controller/src/beerocks/master/son_management.cpp:
```diff
@@ -1917,7 +1917,10 @@ void son_management::handle_bml_message(Socket *sd,
             auto channel_pool_set =
                 std::unordered_set<uint8_t>(channel_pool, channel_pool + channel_pool_size);
-            op_error_code = database.set_channel_scan_pool(radio_mac, channel_pool_set, false)
+            auto scan_all_channels =
```
So ugly...

itayx replied:
Updated the code, and moved the logic to a helper function in the son_management

tomereli reviewed (approved):
Approving cause its a bug fix, but I think this is a very bad change and we should consider splitting the API for better readability.

itayx committed:
controller: son_management: add support for all-channel scan.

The "all-channels" scan uses a specific channel pool to be used in order
to denote that the user wishes to scan on all channel.
The pool is a set of only a single value set to "0".
When the pool matches the requirements for an "all-channel" scan, all
the radio's supported channels, need to be set as the DCS's channel
pool.

Add function that identifies the "all-channel" pool.
Add function that gets the radio's support channels and sets them to the
given channel pool.

Signed-off-by: Itay Elenzweig <itayx.elenzweig@intel.com>

adam1985d commented on controller/src/beerocks/master/tasks/dynamic_channel_selection_task.cpp:
```diff
@@ -52,54 +52,42 @@ beerocks::eChannelScanStatusCode dynamic_channel_selection_task::dcs_request_sca
     // get the parent node to send the CMDU to the agent
-    auto ire = database.get_node_parent_ire(radio_mac_str);
+    auto radio_mac_str = tlvf::mac_to_string(m_radio_mac);
+    auto ire           = database.get_node_parent_ire(radio_mac_str);
```
need to add a check here:
if we have for some reason `get_node_parent_ire` return an empty string - we don't check this and perform `send_cmdu_to_agent_ with empty std::string

itayx replied:
Added check, in case `ire` returns as an empty string the DCS will reach the `Abort` state with an `INTERNAL_FAILURE` status

adam1985d commented on controller/src/beerocks/master/son_management.cpp:
```diff
@@ -30,6 +30,34 @@ using namespace beerocks;
+ * @param database:          Reference to the DB.
+ * @param radio_mac:         MAC address of radio.
+ */
+static void set_pool_to_all_supported_channels(std::unordered_set<uint8_t> &channel_pool_set,
```
instead of `set_pool_to_all_supported_channels` I would suggest a `get_pool_of_all_supported_channels` which is what you do.

itayx replied:
Agree, much better name,
Changed

adam1985d commented on controller/src/beerocks/master/son_management.cpp:
```diff
@@ -30,6 +30,34 @@ using namespace beerocks;
+{
+    LOG(DEBUG) << "Setting channel pool to all channels";
+    channel_pool_set.clear();
+    auto all_channels = database.get_hostap_supported_channels(tlvf::mac_to_string(radio_mac));
```
`get_hostap_supported_channels` can fail and return an empty string `std::vector<beerocks::message::sWifiChannel>()`
please add a check here.

this means the function should return `bool` and not `void`

itayx replied:
Even if `get_hostap_supported_channels` fails and we get an empty `std::vector<beerocks::message::sWifiChannel>()` it won't matter since later on in the DCS task we validate the pool with
```
if (curr_channel_pool.empty()) {
        LOG(ERROR) << "empty channel pool is not supported. please set channel pool for mac="
                   << m_radio_mac;
        return beerocks::eChannelScanStatusCode::INVALID_PARAMS;
}
```
So even if an empty pool is set, the scan will fail with an `INVALID_PARAMS` in `m_last_scan_error_code`

adam1985d replied:
we shouldn't care - the DCS task should not depend on the way we decide to implement the son_management and vise versa - or to be more precise - the dependency on the implementation should be minimal - each one should do its own part without depending on the other to verify for it.

son_management should verify it doesn't set and invalid parameter and then responds to the user "all params are ok" and the task should not send a scan request with invalid params.

itayx replied:
You're right,
Updated

adam1985d commented on controller/src/beerocks/master/son_management.cpp:
```diff
@@ -1950,8 +1978,12 @@ void son_management::handle_bml_message(Socket *sd,
+            auto u_channel_pool_size = uint8_t(channel_pool_size);
             auto channel_pool_set =
                 std::unordered_set<uint8_t>(channel_pool, channel_pool + channel_pool_size);
+            if (channel_pool_size == 1 && *channel_pool_set.begin() == SCAN_ALL_CHANNELS)
```
I understand why @tomereli  doesn't like the `*channel_pool_set.begin() == SCAN_ALL_CHANNELS)`
if we are going to "learn to live with it" - at least wrap it with a meaningful naming function:
```
bool is_scan_all_channels_request(channel_pool_set_type &channel_pool_set){
    return ((channel_pool_set == 1) &&*channel_pool_set.begin() == SCAN_ALL_CHANNELS));
}
```
or something similar

itayx replied:
The helper function actually makes the code look much better.
Changed

adam1985d reviewed (approved):
looks much better and cleaner now that you moved it to the son_managemet.
please fix my comments - approving

adam1985d reviewed (changes requested):
please see my comment on the last commit

adam1985d commented on controller/src/beerocks/master/son_management.cpp:
```diff
@@ -1952,6 +1994,11 @@ void son_management::handle_bml_message(Socket *sd,
             op_error_code == eChannelScanOperationCode::SUCCESS) {
             auto channel_pool_set =
                 std::unordered_set<uint8_t>(channel_pool, channel_pool + channel_pool_size);
+            if (is_scan_all_channels_request(channel_pool_set))
```
1. you didn't add `{}` which is against our coding convention
2. in case of error you are still setting the DB! this part should be fixed similar to what you did in the `beerocks_message::ACTION_BML_CHANNEL_SCAN_START_SCAN_REQUEST` handling
